protocol-units/da-sequencer/node/src/storage/mod.rs

```diff
@@ -1,74 +1,404 @@
+
+		Ok(Storage { db: Arc::new(db) })
+	}
+
+	pub fn write_batch(&self, batch: DaBatch<FullNodeTx>) -> Result<(), DaSequencerError> {
+		let cf = self.db.cf_handle(cf::PENDING_TRANSACTIONS).ok_or_else(|| {
+			DaSequencerError::Generic("Missing column family: pending_transactions".into())
+		})?;
+		let tx = batch.data();
+		let key = tx.id.0;
+		let value = bcs::to_bytes(tx)
+			.map_err(|e| DaSequencerError::Generic(format!("Serialization error: {}", e)))?;
+
+		let mut write_batch = WriteBatch::default();
+		write_batch.put_cf(&cf, key, value);
+
+		self.db
+			.write(write_batch)
+			.map_err(|e| DaSequencerError::Generic(format!("DB write error: {}", e)))?;
+
+		Ok(())
 	}
 
-	/// Return, if exists, the Block at the given height.
 	pub fn get_block_at_height(
 		&self,
 		height: BlockHeight,
-	) -> std::result::Result<Option<SequencerBlock>, DaSequencerError> {
-		todo!();
+	) -> Result<Option<SequencerBlock>, DaSequencerError> {
+		let cf = self
+			.db
+			.cf_handle(cf::BLOCKS)
+			.ok_or_else(|| DaSequencerError::Generic("Missing column family: blocks".into()))?;
+
+		let key: [u8; 8] = height.0.to_be_bytes();
+
+		match self.db.get_cf(&cf, key).map_err(|e| DaSequencerError::Generic(e.to_string()))? {
+			Some(bytes) => {
+				let block: SequencerBlock = bcs::from_bytes(&bytes).map_err(|e| {
+					DaSequencerError::Generic(format!("Deserialization error: {}", e))
+				})?;
+				Ok(Some(block))
+			}
+			None => Ok(None),
+		}
 	}
 
-	/// Return, if exists, the Block with specified sequencer id.
 	pub fn get_block_with_digest(
 		&self,
 		id: SequencerBlockDigest,
-	) -> std::result::Result<Option<SequencerBlock>, DaSequencerError> {
-		todo!();
+	) -> Result<Option<SequencerBlock>, DaSequencerError> {
+		let cf = self.db.cf_handle(cf::BLOCKS_BY_DIGEST).ok_or_else(|| {
+			DaSequencerError::Generic("Missing column family: blocks_by_digest".into())
+		})?;
+
+		let key = id.0;
+
+		let height_bytes =
+			match self.db.get_cf(&cf, key).map_err(|e| DaSequencerError::Generic(e.to_string()))? {
+				Some(bytes) => bytes,
+				None => return Ok(None),
+			};
+
+		if height_bytes.len() != 8 {
+			return Err(DaSequencerError::Generic(
+				"Invalid height length in digest mapping".into(),
+			));
+		}
+
+		let mut arr = [0u8; 8];
+		arr.copy_from_slice(&height_bytes);
+		let height = BlockHeight(u64::from_be_bytes(arr));
+
+		self.get_block_at_height(height)
 	}
 
-	/// Produce next block with pending Tx.
-	/// Generate the new height.
-	/// Aggregate all pending Tx until the block is filled.
-	/// A block is filled if no more Tx are pending or it's size is more than the max size.
-	/// All pending Tx added to the block are removed from the pending Tx table.
-	/// Save the block for this height
-	/// Return the block.
-	pub fn produce_next_block(
-		&self,
-	) -> std::result::Result<Option<SequencerBlock>, DaSequencerError> {
-		todo!();
+	pub fn produce_next_block(&self) -> Result<Option<SequencerBlock>, DaSequencerError> {
```

Comment:
This part should be rethought, we speak about it tomorrow. In the first review, I didn't see this file (it was hidden) so I miss it. I thought the block part hasn't been done.

Reply:
To be addressed in followup `produce_block` PR.